lib/connection_controller.go

```diff
@@ -0,0 +1,330 @@
+	// we need to connect to a new outbound peer, it chooses one of the addresses
+	// it's aware of at random and provides it to us.
+	AddrMgr *addrmgr.AddrManager
+
+	// When --connectips is set, we don't connect to anything from the addrmgr.
+	connectIps []string
+
+	// The target number of outbound peers we want to have.
+	targetOutboundPeers uint32
+	// The maximum number of inbound peers we allow.
+	maxInboundPeers uint32
+	// When true, only one connection per IP is allowed. Prevents eclipse attacks
+	// among other things.
+	limitOneInboundConnectionPerIP bool
+}
+
+func NewConnectionController(params *DeSoParams, cmgr *ConnectionManager, handshakeController *HandshakeController,
+	rnManager *RemoteNodeManager, blsKeystore *BLSKeystore, addrMgr *addrmgr.AddrManager, targetOutboundPeers uint32,
+	maxInboundPeers uint32, limitOneInboundConnectionPerIP bool) *ConnectionController {
+
+	return &ConnectionController{
+		params:                         params,
+		cmgr:                           cmgr,
+		blsKeystore:                    blsKeystore,
+		handshake:                      handshakeController,
+		rnManager:                      rnManager,
+		AddrMgr:                        addrMgr,
+		targetOutboundPeers:            targetOutboundPeers,
+		maxInboundPeers:                maxInboundPeers,
+		limitOneInboundConnectionPerIP: limitOneInboundConnectionPerIP,
+	}
+}
+
+func (cc *ConnectionController) GetRemoteNodeManager() *RemoteNodeManager {
+	return cc.rnManager
+}
+
+// ###########################
+// ## Handlers (Peer, DeSoMessage)
+// ###########################
+
+func (cc *ConnectionController) _handleDonePeerMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypeDisconnectedPeer {
+		return
+	}
+
+	cc.rnManager.DisconnectById(NewRemoteNodeId(origin.ID))
+}
+
+func (cc *ConnectionController) _handleAddrMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypeAddr {
+		return
+	}
+
+	// TODO
+}
+
+func (cc *ConnectionController) _handleGetAddrMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypeGetAddr {
+		return
+	}
+
+	// TODO
+}
+
+// _handleNewConnectionMessage is called when a new outbound or inbound connection is established. It is responsible
+// for creating a RemoteNode from the connection and initiating the handshake.
+func (cc *ConnectionController) _handleNewConnectionMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypeNewConnection {
+		return
+	}
+
+	var msg *MsgDeSoNewConnection
+	var ok bool
+	if msg, ok = desoMsg.(*MsgDeSoNewConnection); !ok {
+		return
+	}
+
+	var remoteNode *RemoteNode
+	var err error
+	switch msg.Connection.GetConnectionType() {
+	case ConnectionTypeInbound:
+		remoteNode, err = cc.processInboundConnection(msg.Connection)
+		if err != nil {
+			glog.Errorf("ConnectionController.handleNewConnectionMessage: Problem handling inbound connection: %v", err)
+			msg.Connection.Close()
+			return
```

Comment:
I modified this function slightly in the next PR. Generally, the connection error should be handled within the context of this function. If we returned an error here, it would mean we should quit the node, but actually, the error is likely due to a misbehaving peer. In this case, our desired action is to close the connection.